File: java/bowling/bowling_180607.java
public class BowlingGame {
    
    private int currentRoll = 0;
    private ArrayList<Integer> rolls = new ArrayList<Integer>();

    public void roll(int pins) {
        this.rolls.add(pins);
    }

    public int score() {
        int currentScore = 0;
        int frameIndex = 0;
        for (int frame = 0; frame < 10; frame ++) {
            if (this.isSpare(frameIndex)) {
                currentScore += (10 + this.spareBonus(frameIndex));
                frameIndex += 2;
            } else if (this.isStrike(frameIndex)) {
                currentScore += (10 + this.strikeBonus(frameIndex));
                frameIndex += 1;
            } else {
                currentScore += this.sumOfPins(frameIndex);
                frameIndex += 2;
            }
        }
    }

    private boolean isSpare(int frameIndex) {
        return (this.rolls[frameIndex] + this.rolls[frameIndex + 1]) == 10;
    }

    private boolean isStrike(int frameIndex) {
        return this.rolls[frameIndex] == 10
    }

    private int spareBonus(int frameIndex) {
        return this.rolls[frameIndex + 2];
    }

    private int strikeBonus(int frameIndex) {
        return this.rolls[frameIndex + 1] + this.rolls[frameIndex + 2]
    }

    private int sumOfPins(int frameIndex) {
        return this.rolls[frameIndex] + this.rolls[frameIndex + 1];
    }
}

File: javascript/bowling/180607/bowling_180607.js
export default class Bowling {

    constructor() {
        this.rolls = []
    }

    roll(pins) {
        this.rolls.push(pins);
    }

    isStrike(index) {
        return (this.rolls[index] == 10);
    }

    isSpare(index) {
        return (this.rolls[index] + this.rolls[index + 1] == 10);
    }

    strikeBonus(index) {
        return 10 + this.rolls[index + 1] + this.rolls[index + 2];
    }

    spareBonus(index) {
        return 10 + this.rolls[index + 2];
    }

    score() {
        let totalScore = 0;
        let index = 0;
        for (let frame = 0; frame < 10; frame ++) {
            if (this.isStrike(index)) {
                totalScore += this.strikeBonus(index);
                index += 1;
            } else if (this.isSpare(index)) {
                totalScore += this.spareBonus(index);
                index += 2;
            } else {
                totalScore += (this.rolls[index] + this.rolls[index + 1]);
                index += 2;
            }
        }
        return totalScore;
    }
}